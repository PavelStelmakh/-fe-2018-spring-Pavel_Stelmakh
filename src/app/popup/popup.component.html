<div [class.showPopup]="show">
  <ng-content></ng-content>  
</div>
