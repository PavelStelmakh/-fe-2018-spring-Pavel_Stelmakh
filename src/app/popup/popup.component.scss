:host {
    display: block;
    div {
        position: absolute;
        width: 300px;
        height: 70px;
        background-color: rgba(102, 102, 102, 0.8);
        border: 1px solid #666;
        border-left-width: 10px;
        top: -72px;
        left: calc(50vw - 150px);
        transition: top 500ms linear;
        line-height: 70px;
        font-family: Verdana;
        text-align: center;
        color: #ccc;
        font-size: 1.1rem;
    }

    .showPopup {
        top: 0;
    }

}
