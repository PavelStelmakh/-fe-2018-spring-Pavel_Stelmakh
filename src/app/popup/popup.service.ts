import { Injectable, OnDestroy } from '@angular/core';
import { BehaviorSubject } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class PopupService implements OnDestroy {
  showPopup: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);

  constructor() { }

  show() {
    this.showPopup.next(true);
   setTimeout(() => this.showPopup.next(false), 3000);
  }

  ngOnDestroy() {
    this.showPopup.complete();
  }

}
