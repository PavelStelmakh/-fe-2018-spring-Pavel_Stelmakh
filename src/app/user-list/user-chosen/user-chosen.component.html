<div [class.chosen]="true">
  <div>
    <div>
      <span>{{'user-details.output-form.name' | translate}}:</span>
      <span>{{user?.name}}</span>
    </div>
    <div>
        <span>{{'user-details.output-form.age' | translate}}:</span>
        <span>{{user?.age}}</span>
    </div>
    <div>
      <span>{{'user-details.output-form.information' | translate}}:</span>
      <span>{{user?.information}}</span>
    </div>
  </div>
  <div [class.expand]="expand" (click)="expandList()">
    <fa-icon [icon]="faCaretDown" size="3x" style="color: #ccc"></fa-icon>
  </div>
</div>
<div [class.buttons]="true">
  <button (click)="deleteUser()" [disabled]="!user">{{'user-details.user-chosen.DeleteButton' | translate}}</button>
  <button (click)="editUser()" [disabled]="!user">{{'user-details.user-chosen.EditButton' | translate}}</button>
  <button (click)="addUser()">{{'user-details.user-chosen.AddButton' | translate}}</button>
</div>
