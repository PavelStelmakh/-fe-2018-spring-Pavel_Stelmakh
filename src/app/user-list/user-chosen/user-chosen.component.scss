.chosen {
    width: 500px;
    display: flex;
    justify-content: space-around;
    text-align: start;

    & > div:first-child {
        width: 220px;

        & > div {
            display: flex;
            justify-content: space-between;
            & > span {
                white-space: nowrap;
                overflow: hidden;
                text-overflow: ellipsis;
            }
        }

    }

    & > div:last-child {
        display: inline-block;
        transform: rotate(90deg);
        transition: transform 100ms linear;
    }

    & > div.expand {
        transform: rotate(0deg);
    }

}

