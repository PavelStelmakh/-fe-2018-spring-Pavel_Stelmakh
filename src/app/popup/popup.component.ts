import { Component, OnInit } from '@angular/core';
import { PopupService } from './popup.service';

@Component({
  selector: 'app-popup',
  templateUrl: './popup.component.html',
  styleUrls: ['./popup.component.scss']
})
export class PopupComponent implements OnInit {
  show: boolean;

  constructor(private popup: PopupService) {}

  ngOnInit() {
    this.show = false;
    this.popup.showPopup.subscribe((result: boolean) => this.show = result);
  }

}
