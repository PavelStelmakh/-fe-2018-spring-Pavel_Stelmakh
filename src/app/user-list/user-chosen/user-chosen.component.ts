import { Component, Input, Output, EventEmitter, OnInit } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { User } from 'models/User';
import { faCaretDown, IconDefinition } from '@fortawesome/free-solid-svg-icons';
import { PopupService } from 'src/app/popup/popup.service';
import * as usersAction from '../../actions/users.action';
import * as usersReducer from '../../reducers/users.reducer';

@Component({
  selector: 'app-user-chosen',
  templateUrl: './user-chosen.component.html',
  styleUrls: ['./user-chosen.component.scss']
})
export class UserChosenComponent implements OnInit {
  user: User;
  faCaretDown: IconDefinition = faCaretDown;
  expand: boolean;
  @Output('expand') expandEvent: EventEmitter<boolean> = new EventEmitter<boolean>();
  @Output() delete: EventEmitter<void> = new EventEmitter<void>();
  @Output() edit: EventEmitter<void> = new EventEmitter<void>();

  constructor(
    private popup: PopupService,
    private store: Store<usersReducer.State>
  ) {}

  ngOnInit() {
    this.expand = false;
    this.expandEvent.emit(this.expand);
    this.store.pipe(select(usersReducer.getSelectedUser)).subscribe((user: User) => this.user = user);
  }

  expandList() {
    this.expand = !this.expand;
    this.expandEvent.emit(this.expand);
  }

  deleteUser() {
    this.store.dispatch(new usersAction.DeleteUserAction(this.user.id));
    this.user = null;
    this.popup.show();
    this.delete.emit();
  }

  editUser() {    
    this.store.dispatch(new usersAction.SelectUserAction(this.user.id));
    this.edit.emit();
  }

  addUser() {
    this.store.dispatch(new usersAction.SelectUserAction(null));
    this.edit.emit();
  }

}
