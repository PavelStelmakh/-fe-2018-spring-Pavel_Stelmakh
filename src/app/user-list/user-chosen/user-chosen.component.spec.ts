import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { UserChosenComponent } from './user-chosen.component';
import { Observable, of } from 'rxjs';
import { TranslateModule, TranslateLoader, TranslateService } from '@ngx-translate/core';
import { Store } from '@ngrx/store';
import * as profileReducer from '../../reducers/profile.reduser';
import { PopupService } from '../../popup/popup.service';

const translations = {
  "user-details": {
    "output-form": {
      "name": "Name",
      "password": "Password",
      "age": "Age",
      "birthday": "Birthday",
      "dateOfLogin": "Date of login",
      "information": "Information",
      "dateOfNotification": "Date of notification"
    },
    "user-chosen": {
        "AddButton": "Add",
        "DeleteButton": "Delete",
        "EditButton": "Edit"
    }
  }
};

class FakeLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<{}> {
    return of(translations);
  }
}

describe('UserChosenComponent', () => {
  let fixture: ComponentFixture<UserChosenComponent>;
  let translate: TranslateService;
  let store: Store<profileReducer.State>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        TranslateModule.forRoot({
          loader: {provide: TranslateLoader, useClass: FakeLoader}
        })
      ],
      declarations: [
        UserChosenComponent
      ],
      providers: [
        PopupService,
        {
          provide: Store,
          useValue: {
            dispatch: jest.fn(),
            pipe: jest.fn(() => of({}))
          }
        }
      ],
      schemas: [
       NO_ERRORS_SCHEMA
      ]
    }).compileComponents();

  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserChosenComponent);
    translate = TestBed.get(TranslateService);
    store = TestBed.get(Store);
  });

  it('renders markup to snapshot', () => {
    translate.use('en');
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

});
