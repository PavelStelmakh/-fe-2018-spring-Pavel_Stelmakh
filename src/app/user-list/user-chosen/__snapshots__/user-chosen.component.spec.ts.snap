// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserChosenComponent renders markup to snapshot 1`] = `
<app-user-chosen
  delete={[Function EventEmitter]}
  edit={[Function EventEmitter]}
  expand="false"
  expandEvent={[Function EventEmitter]}
  faCaretDown={[Function Object]}
  popup={[Function PopupService]}
  store={[Function Object]}
  user={[Function Object]}
>
  <div
    class="chosen"
  >
    <div>
      <div>
        <span>
          Name:
        </span>
        <span>
          
        </span>
      </div>
      <div>
        <span>
          Age:
        </span>
        <span>
          
        </span>
      </div>
      <div>
        <span>
          Information:
        </span>
        <span>
          
        </span>
      </div>
    </div>
    <div>
      <fa-icon
        size="3x"
        style="color: #ccc"
      />
    </div>
  </div><div
    class="buttons"
  >
    <button>
      Delete
    </button>
    <button>
      Edit
    </button>
    <button>
      Add
    </button>
  </div>
</app-user-chosen>
`;
